addons/account/report/account_invoice_report.py

```diff
@@ -188,12 +188,37 @@ def get_rate(currency_id):
                 self.env.company,
                 self._fields['invoice_date'].today()
             )
-        result = super(AccountInvoiceReport, self).read_group(domain, fields, set(groupby) | {'currency_id'}, offset, limit, orderby, lazy)
+
+        # First we get the structure of the results. The results won't be correct in multi-currency,
+        # but we need this result structure.
+        # By adding 'ids:array_agg(id)' to the fields, we will be able to map the results of the
+        # second step in the structure of the first step.
+        result_ref = super(AccountInvoiceReport, self).read_group(
+            domain, fields + ['ids:array_agg(id)'], groupby, offset, limit, orderby, lazy
+        )
+        for res_ref in result_ref:
+            for field in {'amount_total', 'price_average', 'price_subtotal', 'residual'} & set(res_ref):
+                res_ref[field] = 0.0
+
+        # Then we perform another read_group, but this time we group by 'currency_id'. This way, we
+        # are able to convert in batch in the current company currency.
+        # During the process, we fill in the result structure we got in the previous step. To make
+        # the mapping, we use the aggregated ids.
+        # This is O(n²), but it shouldn't really a problem in comparison to the time taken for the
```

Comment:
Isn't it in `O(2n)` ? You do 2 `read_group` instead of one, one with and one without the `GROUP BY currency_id` in SQL which doesn't change the complexity.
Or where you talking about the nested for loops below? In that case, it is more `O(n.m)` with `n` the number of lines and `m` the number of currencies (which will always be small / limited). 
Am I getting something wrong?

Reply:
Indeed I was referring to the nested loops. You are right it's `O(n.m)` since the lists have a different sizes. But indeed, I don't think this would cause a severe performance drop, this is just loop on some lists with a reasonable size.